File: services/go/internal/capingest/parse.go
package capingest

import (
	"bytes"
	"encoding/xml"
	"fmt"
	"io"
	"net/url"
	"strings"
	"time"
)

type capXML struct {
	XMLName    xml.Name  `xml:"alert"`
	Identifier string    `xml:"identifier"`
	Sender     string    `xml:"sender"`
	Sent       string    `xml:"sent"`
	Status     string    `xml:"status"`
	MsgType    string    `xml:"msgType"`
	Scope      string    `xml:"scope"`
	Note       string    `xml:"note"`
	Code       []string  `xml:"code"`
	References string    `xml:"references"`
	Incidents  string    `xml:"incidents"`
	Infos      []infoXML `xml:"info"`
}

type infoXML struct {
	Language    string        `xml:"language"`
	Category    []string      `xml:"category"`
	Event       string        `xml:"event"`
	Response    []string      `xml:"responseType"`
	Urgency     string        `xml:"urgency"`
	Severity    string        `xml:"severity"`
	Certainty   string        `xml:"certainty"`
	Audience    string        `xml:"audience"`
	Effective   string        `xml:"effective"`
	Onset       string        `xml:"onset"`
	Expires     string        `xml:"expires"`
	SenderName  string        `xml:"senderName"`
	Headline    string        `xml:"headline"`
	Description string        `xml:"description"`
	Instruction string        `xml:"instruction"`
	Web         string        `xml:"web"`
	EventCodes  []pairXML     `xml:"eventCode"`
	Areas       []areaXML     `xml:"area"`
	Parameters  []pairXML     `xml:"parameter"`
	Resources   []resourceXML `xml:"resource"`
}

type areaXML struct {
	Description string    `xml:"areaDesc"`
	Polygons    []string  `xml:"polygon"`
	Circles     []string  `xml:"circle"`
	Geocodes    []pairXML `xml:"geocode"`
}

type pairXML struct {
	Name  string `xml:"valueName"`
	Value string `xml:"value"`
}

type resourceXML struct {
	Description string `xml:"resourceDesc"`
	MimeType    string `xml:"mimeType"`
	URI         string `xml:"uri"`
	DerefURI    string `xml:"derefUri"`
}

// ParseCAP parses a CAP XML document into a normalized alert.
func ParseCAP(raw []byte) (Alert, error) {
	decoder := xml.NewDecoder(bytes.NewReader(raw))
	decoder.Strict = true
	var parsed capXML
	if err := decoder.Decode(&parsed); err != nil {
		return Alert{}, err
	}
	if parsed.XMLName.Local != "alert" {
		return Alert{}, fmt.Errorf("CAP root element is %q, expected alert", parsed.XMLName.Local)
	}

	alert := Alert{
		Identifier:  clean(parsed.Identifier),
		Sender:      clean(parsed.Sender),
		Sent:        clean(parsed.Sent),
		Status:      clean(parsed.Status),
		MessageType: clean(parsed.MsgType),
		Scope:       clean(parsed.Scope),
		Note:        clean(parsed.Note),
		Code:        cleanSlice(parsed.Code),
		References:  clean(parsed.References),
		Incidents:   clean(parsed.Incidents),
		RawXML:      string(raw),
	}
	for _, info := range parsed.Infos {
		alert.Infos = append(alert.Infos, normalizeInfo(info))
	}
	alert.Warnings = validateCAP(alert)
	if len(alert.Warnings) > 0 && hasFatalCAPWarning(alert.Warnings) {
		return Alert{}, fmt.Errorf("invalid CAP alert %s: %s", alert.Identifier, strings.Join(alert.Warnings, "; "))
	}
	return alert, nil
}

func validateCAP(alert Alert) []string {
	var warnings []string
	required := []struct {
		name  string
		value string
	}{
		{"identifier", alert.Identifier},
		{"sender", alert.Sender},
		{"sent", alert.Sent},
		{"status", alert.Status},
		{"msgType", alert.MessageType},
		{"scope", alert.Scope},
	}
	for _, item := range required {
		if strings.TrimSpace(item.value) == "" {
			warnings = append(warnings, "fatal: missing "+item.name)
		}
	}
	if alert.Sent != "" && parseCAPTime(alert.Sent).IsZero() {
		warnings = append(warnings, "fatal: invalid sent timestamp")
	}
	warnings = appendEnumWarning(warnings, "status", alert.Status, []string{"Actual", "Exercise", "System", "Test", "Draft"}, true)
	warnings = appendEnumWarning(warnings, "msgType", alert.MessageType, []string{"Alert", "Update", "Cancel", "Ack", "Error"}, true)
	warnings = appendEnumWarning(warnings, "scope", alert.Scope, []string{"Public", "Restricted", "Private"}, true)
	if len(alert.Infos) == 0 && !strings.EqualFold(alert.MessageType, "Cancel") {
		warnings = append(warnings, "fatal: non-cancel alert has no info block")
	}
	for index, info := range alert.Infos {
		prefix := fmt.Sprintf("info[%d]", index)
		if info.Event == "" {
			warnings = append(warnings, prefix+": missing event")
		}
		warnings = appendEnumWarning(warnings, prefix+".urgency", info.Urgency, []string{"Immediate", "Expected", "Future", "Past", "Unknown"}, false)
		warnings = appendEnumWarning(warnings, prefix+".severity", info.Severity, []string{"Extreme", "Severe", "Moderate", "Minor", "Unknown"}, false)
		warnings = appendEnumWarning(warnings, prefix+".certainty", info.Certainty, []string{"Observed", "Likely", "Possible", "Unlikely", "Unknown"}, false)
		for _, raw := range []struct {
			name  string
			value string
		}{
			{"effective", info.Effective},
			{"onset", info.Onset},
			{"expires", info.Expires},
		} {
			if raw.value != "" && parseCAPTime(raw.value).IsZero() {
				warnings = append(warnings, prefix+": invalid "+raw.name+" timestamp")
			}
		}
		for _, resource := range info.Resources {
			if resource.URI != "" {
				if parsed, err := url.Parse(resource.URI); err != nil || parsed.Scheme == "" && !strings.HasPrefix(resource.URI, "cid:") {
					warnings = append(warnings, prefix+": resource URI is not absolute")
				}
			}
		}
	}
	return warnings
}

func appendEnumWarning(warnings []string, name string, value string, allowed []string, fatal bool) []string {
	if strings.TrimSpace(value) == "" {
		return warnings
	}
	for _, item := range allowed {
		if strings.EqualFold(value, item) {
			return warnings
		}
	}
	prefix := ""
	if fatal {
		prefix = "fatal: "
	}
	return append(warnings, prefix+"invalid "+name+" "+value)
}

func hasFatalCAPWarning(warnings []string) bool {
	for _, warning := range warnings {
		if strings.HasPrefix(warning, "fatal:") {
			return true
		}
	}
	return false
}

func parseCAPTime(raw string) time.Time {
	if parsed, err := time.Parse(time.RFC3339Nano, strings.TrimSpace(raw)); err == nil {
		return parsed
	}
	return time.Time{}
}

// ParseAtomEntries extracts CAP links from an Atom feed.
func ParseAtomEntries(raw []byte) ([]AtomEntry, error) {
	decoder := xml.NewDecoder(bytes.NewReader(raw))
	var entries []AtomEntry
	var current *AtomEntry

	for {
		token, err := decoder.Token()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, err
		}

		switch tok := token.(type) {
		case xml.StartElement:
			switch tok.Name.Local {
			case "entry":
				current = &AtomEntry{}
			case "id":
				if current != nil {
					current.ID = readElementText(decoder, tok)
				}
			case "updated":
				if current != nil {
					current.Updated = readElementText(decoder, tok)
				}
			case "link":
				if current != nil {
					if href := linkHref(tok); href != "" {
						current.Links = appendCAPLink(current.Links, href)
					}
				}
			}
		case xml.EndElement:
			if tok.Name.Local == "entry" && current != nil {
				if current.ID != "" && len(current.Links) == 0 && strings.HasPrefix(current.ID, "http") {
					current.Links = appendCAPLink(current.Links, current.ID)
				}
				if current.ID != "" && len(current.Links) > 0 {
					entries = append(entries, *current)
				}
				current = nil
			}
		}
	}

	return entries, nil
}

func normalizeInfo(info infoXML) AlertInfo {
	result := AlertInfo{
		Language:    clean(info.Language),
		Category:    cleanSlice(info.Category),
		Event:       clean(info.Event),
		Response:    cleanSlice(info.Response),
		Urgency:     clean(info.Urgency),
		Severity:    clean(info.Severity),
		Certainty:   clean(info.Certainty),
		Audience:    clean(info.Audience),
		Effective:   clean(info.Effective),
		Onset:       clean(info.Onset),
		Expires:     clean(info.Expires),
		SenderName:  clean(info.SenderName),
		Headline:    clean(info.Headline),
		Description: clean(info.Description),
		Instruction: clean(info.Instruction),
		Web:         clean(info.Web),
		EventCodes:  normalizePairs(info.EventCodes),
	}
	for _, area := range info.Areas {
		result.Areas = append(result.Areas, AlertArea{
			Description: clean(area.Description),
			Polygons:    cleanSlice(area.Polygons),
			Circles:     cleanSlice(area.Circles),
			Geocodes:    normalizePairs(area.Geocodes),
		})
	}
	result.Parameters = normalizePairs(info.Parameters)
	for _, resource := range info.Resources {
		result.Resources = append(result.Resources, Resource{
			Description: clean(resource.Description),
			MimeType:    clean(resource.MimeType),
			URI:         clean(resource.URI),
			DerefURI:    clean(resource.DerefURI),
		})
	}
	return result
}

func normalizePairs(values []pairXML) []NameValue {
	pairs := make([]NameValue, 0, len(values))
	for _, value := range values {
		name := clean(value.Name)
		item := clean(value.Value)
		if name == "" && item == "" {
			continue
		}
		pairs = append(pairs, NameValue{Name: name, Value: item})
	}
	return pairs
}

func readElementText(decoder *xml.Decoder, start xml.StartElement) string {
	var value string
	if err := decoder.DecodeElement(&value, &start); err != nil {
		return ""
	}
	return clean(value)
}

func linkHref(start xml.StartElement) string {
	for _, attr := range start.Attr {
		if attr.Name.Local == "href" {
			return clean(attr.Value)
		}
	}
	return ""
}

func appendCAPLink(links []string, href string) []string {
	if href == "" {
		return links
	}
	for _, link := range links {
		if link == href {
			return links
		}
	}
	links = append(links, href)
	if strings.HasPrefix(href, "http") && !strings.HasSuffix(href, ".cap") {
		capURL := strings.TrimRight(href, "/") + ".cap"
		for _, link := range links {
			if link == capURL {
				return links
			}
		}
		links = append(links, capURL)
	}
	return links
}

func cleanSlice(values []string) []string {
	result := make([]string, 0, len(values))
	for _, value := range values {
		cleaned := clean(value)
		if cleaned != "" {
			result = append(result, cleaned)
		}
	}
	return result
}

func clean(value string) string {
	return strings.TrimSpace(value)
}

File: crates/haze-cap/src/parse.rs
use std::collections::HashSet;

use chrono::{DateTime, FixedOffset};
use quick_xml::de::from_str;
use serde::Deserialize;
use thiserror::Error;
use url::Url;

use crate::model::{Alert, AlertArea, AlertInfo, AtomEntry, NameValue, Resource};

#[derive(Debug, Error)]
pub enum ParseError {
    #[error("CAP XML is not UTF-8")]
    Utf8(#[from] std::str::Utf8Error),
    #[error("failed to parse XML: {0}")]
    Xml(#[from] quick_xml::DeError),
    #[error("invalid CAP alert {identifier}: {warnings}")]
    InvalidCap {
        identifier: String,
        warnings: String,
    },
}

#[derive(Debug, Default, Deserialize)]
#[serde(rename = "alert")]
struct CapXml {
    #[serde(default)]
    identifier: String,
    #[serde(default)]
    sender: String,
    #[serde(default)]
    sent: String,
    #[serde(default)]
    status: String,
    #[serde(rename = "msgType", default)]
    msg_type: String,
    #[serde(default)]
    scope: String,
    #[serde(default)]
    note: String,
    #[serde(rename = "code", default)]
    code: Vec<String>,
    #[serde(default)]
    references: String,
    #[serde(default)]
    incidents: String,
    #[serde(rename = "info", default)]
    infos: Vec<InfoXml>,
}

#[derive(Debug, Default, Deserialize)]
struct InfoXml {
    #[serde(default)]
    language: String,
    #[serde(rename = "category", default)]
    category: Vec<String>,
    #[serde(default)]
    event: String,
    #[serde(rename = "responseType", default)]
    response: Vec<String>,
    #[serde(default)]
    urgency: String,
    #[serde(default)]
    severity: String,
    #[serde(default)]
    certainty: String,
    #[serde(default)]
    audience: String,
    #[serde(default)]
    effective: String,
    #[serde(default)]
    onset: String,
    #[serde(default)]
    expires: String,
    #[serde(rename = "senderName", default)]
    sender_name: String,
    #[serde(default)]
    headline: String,
    #[serde(default)]
    description: String,
    #[serde(default)]
    instruction: String,
    #[serde(default)]
    web: String,
    #[serde(rename = "eventCode", default)]
    event_codes: Vec<PairXml>,
    #[serde(rename = "area", default)]
    areas: Vec<AreaXml>,
    #[serde(rename = "parameter", default)]
    parameters: Vec<PairXml>,
    #[serde(rename = "resource", default)]
    resources: Vec<ResourceXml>,
}

#[derive(Debug, Default, Deserialize)]
struct AreaXml {
    #[serde(rename = "areaDesc", default)]
    description: String,
    #[serde(rename = "polygon", default)]
    polygons: Vec<String>,
    #[serde(rename = "circle", default)]
    circles: Vec<String>,
    #[serde(rename = "geocode", default)]
    geocodes: Vec<PairXml>,
}

#[derive(Debug, Default, Deserialize)]
struct PairXml {
    #[serde(rename = "valueName", default)]
    name: String,
    #[serde(default)]
    value: String,
}

#[derive(Debug, Default, Deserialize)]
struct ResourceXml {
    #[serde(rename = "resourceDesc", default)]
    description: String,
    #[serde(rename = "mimeType", default)]
    mime_type: String,
    #[serde(default)]
    uri: String,
    #[serde(rename = "derefUri", default)]
    deref_uri: String,
}

#[derive(Debug, Default, Deserialize)]
struct AtomFeedXml {
    #[serde(rename = "entry", default)]
    entries: Vec<AtomEntryXml>,
}

#[derive(Debug, Default, Deserialize)]
struct AtomEntryXml {
    #[serde(default)]
    id: String,
    #[serde(default)]
    updated: String,
    #[serde(rename = "link", default)]
    links: Vec<AtomLinkXml>,
}

#[derive(Debug, Default, Deserialize)]
struct AtomLinkXml {
    #[serde(rename = "@href", default)]
    href: String,
}

pub fn parse_cap(raw: &[u8]) -> Result<Alert, ParseError> {
    let xml = std::str::from_utf8(raw)?;
    let parsed: CapXml = from_str(xml)?;
    let mut alert = Alert {
        identifier: clean(&parsed.identifier),
        sender: clean(&parsed.sender),
        sent: clean(&parsed.sent),
        status: clean(&parsed.status),
        message_type: clean(&parsed.msg_type),
        scope: clean(&parsed.scope),
        note: clean(&parsed.note),
        code: clean_slice(parsed.code),
        references: clean(&parsed.references),
        incidents: clean(&parsed.incidents),
        infos: parsed.infos.into_iter().map(normalize_info).collect(),
        raw_xml: xml.to_string(),
        warnings: Vec::new(),
    };
    alert.warnings = validate_cap(&alert);
    if alert
        .warnings
        .iter()
        .any(|warning| warning.starts_with("fatal:"))
    {
        return Err(ParseError::InvalidCap {
            identifier: alert.identifier.clone(),
            warnings: alert.warnings.join("; "),
        });
    }
    Ok(alert)
}

pub fn parse_atom_entries(raw: &[u8]) -> Result<Vec<AtomEntry>, ParseError> {
    let xml = std::str::from_utf8(raw)?;
    let parsed: AtomFeedXml = from_str(xml)?;
    let mut entries = Vec::new();
    for entry in parsed.entries {
        let id = clean(&entry.id);
        let updated = clean(&entry.updated);
        let mut links = Vec::new();
        for link in entry.links {
            append_cap_link(&mut links, &clean(&link.href));
        }
        if id.starts_with("http") && links.is_empty() {
            append_cap_link(&mut links, &id);
        }
        if !id.is_empty() && !links.is_empty() {
            entries.push(AtomEntry { id, updated, links });
        }
    }
    Ok(entries)
}

fn normalize_info(info: InfoXml) -> AlertInfo {
    AlertInfo {
        language: clean(&info.language),
        category: clean_slice(info.category),
        event: clean(&info.event),
        response: clean_slice(info.response),
        urgency: clean(&info.urgency),
        severity: clean(&info.severity),
        certainty: clean(&info.certainty),
        audience: clean(&info.audience),
        effective: clean(&info.effective),
        onset: clean(&info.onset),
        expires: clean(&info.expires),
        sender_name: clean(&info.sender_name),
        headline: clean(&info.headline),
        description: clean(&info.description),
        instruction: clean(&info.instruction),
        web: clean(&info.web),
        event_codes: normalize_pairs(info.event_codes),
        areas: info
            .areas
            .into_iter()
            .map(|area| AlertArea {
                description: clean(&area.description),
                polygons: clean_slice(area.polygons),
                circles: clean_slice(area.circles),
                geocodes: normalize_pairs(area.geocodes),
            })
            .collect(),
        parameters: normalize_pairs(info.parameters),
        resources: info
            .resources
            .into_iter()
            .map(|resource| Resource {
                description: clean(&resource.description),
                mime_type: clean(&resource.mime_type),
                uri: clean(&resource.uri),
                deref_uri: clean(&resource.deref_uri),
            })
            .collect(),
    }
}

fn validate_cap(alert: &Alert) -> Vec<String> {
    let mut warnings = Vec::new();
    for (name, value) in [
        ("identifier", alert.identifier.as_str()),
        ("sender", alert.sender.as_str()),
        ("sent", alert.sent.as_str()),
        ("status", alert.status.as_str()),
        ("msgType", alert.message_type.as_str()),
        ("scope", alert.scope.as_str()),
    ] {
        if value.trim().is_empty() {
            warnings.push(format!("fatal: missing {name}"));
        }
    }
    if !alert.sent.is_empty() && parse_cap_time(&alert.sent).is_none() {
        warnings.push("fatal: invalid sent timestamp".to_string());
    }
    append_enum_warning(
        &mut warnings,
        "status",
        &alert.status,
        &["Actual", "Exercise", "System", "Test", "Draft"],
        true,
    );
    append_enum_warning(
        &mut warnings,
        "msgType",
        &alert.message_type,
        &["Alert", "Update", "Cancel", "Ack", "Error"],
        true,
    );
    append_enum_warning(
        &mut warnings,
        "scope",
        &alert.scope,
        &["Public", "Restricted", "Private"],
        true,
    );
    if alert.infos.is_empty()
        && !alert.message_type.eq_ignore_ascii_case("Cancel")
        && !alert.status.eq_ignore_ascii_case("System")
    {
        warnings.push("fatal: non-cancel alert has no info block".to_string());
    }
    for (index, info) in alert.infos.iter().enumerate() {
        let prefix = format!("info[{index}]");
        if info.event.is_empty() {
            warnings.push(format!("{prefix}: missing event"));
        }
        append_enum_warning(
            &mut warnings,
            &format!("{prefix}.urgency"),
            &info.urgency,
            &["Immediate", "Expected", "Future", "Past", "Unknown"],
            false,
        );
        append_enum_warning(
            &mut warnings,
            &format!("{prefix}.severity"),
            &info.severity,
            &["Extreme", "Severe", "Moderate", "Minor", "Unknown"],
            false,
        );
        append_enum_warning(
            &mut warnings,
            &format!("{prefix}.certainty"),
            &info.certainty,
            &["Observed", "Likely", "Possible", "Unlikely", "Unknown"],
            false,
        );
        for (name, value) in [
            ("effective", info.effective.as_str()),
            ("onset", info.onset.as_str()),
            ("expires", info.expires.as_str()),
        ] {
            if !value.is_empty() && parse_cap_time(value).is_none() {
                warnings.push(format!("{prefix}: invalid {name} timestamp"));
            }
        }
        for resource in &info.resources {
            if !resource.uri.is_empty() && !resource.uri.starts_with("cid:") {
                match Url::parse(&resource.uri) {
                    Ok(parsed) if !parsed.scheme().is_empty() => {}
                    _ => warnings.push(format!("{prefix}: resource URI is not absolute")),
                }
            }
        }
    }
    warnings
}

fn append_enum_warning(
    warnings: &mut Vec<String>,
    name: &str,
    value: &str,
    allowed: &[&str],
    fatal: bool,
) {
    if value.trim().is_empty() {
        return;
    }
    if allowed.iter().any(|item| value.eq_ignore_ascii_case(item)) {
        return;
    }
    if fatal {
        warnings.push(format!("fatal: invalid {name} {value}"));
    } else {
        warnings.push(format!("invalid {name} {value}"));
    }
}

fn parse_cap_time(raw: &str) -> Option<DateTime<FixedOffset>> {
    DateTime::parse_from_rfc3339(raw.trim()).ok()
}

fn normalize_pairs(values: Vec<PairXml>) -> Vec<NameValue> {
    values
        .into_iter()
        .filter_map(|value| {
            let name = clean(&value.name);
            let value = clean(&value.value);
            if name.is_empty() && value.is_empty() {
                None
            } else {
                Some(NameValue { name, value })
            }
        })
        .collect()
}

fn clean_slice(values: Vec<String>) -> Vec<String> {
    values
        .into_iter()
        .map(|value| clean(&value))
        .filter(|value| !value.is_empty())
        .collect()
}

fn clean(value: &str) -> String {
    value.trim().to_string()
}

fn append_cap_link(links: &mut Vec<String>, href: &str) {
    if href.is_empty() {
        return;
    }
    let mut seen: HashSet<String> = links.iter().cloned().collect();
    if seen.insert(href.to_string()) {
        links.push(href.to_string());
    }
    if href.starts_with("http") && !href.ends_with(".cap") {
        let cap_url = format!("{}.cap", href.trim_end_matches('/'));
        if seen.insert(cap_url.clone()) {
            links.push(cap_url);
        }
    }
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::*;

    #[test]
    fn parses_cap_to_go_compatible_shape() {
        let raw = br#"
<alert xmlns="urn:oasis:names:tc:emergency:cap:1.2">
  <identifier>abc</identifier>
  <sender>sender@example.test</sender>
  <sent>2026-06-22T12:38:00-06:00</sent>
  <status>Actual</status>
  <msgType>Alert</msgType>
  <scope>Public</scope>
  <code>profile:CAP-CP:0.4</code>
  <info>
    <language>en-CA</language>
    <category>Met</category>
    <event>Severe Thunderstorm Warning</event>
    <responseType>Shelter</responseType>
    <urgency>Immediate</urgency>
    <severity>Severe</severity>
    <certainty>Likely</certainty>
    <senderName>Environment Canada</senderName>
    <headline>Severe thunderstorm warning</headline>
    <description>Storm text.</description>
    <instruction>Go indoors.</instruction>
    <eventCode><valueName>SAME</valueName><value>SVR</value></eventCode>
    <area>
      <areaDesc>Saskatoon</areaDesc>
      <geocode><valueName>CLC</valueName><value>065100</value></geocode>
    </area>
    <resource>
      <resourceDesc>audio</resourceDesc>
      <mimeType>audio/mpeg</mimeType>
      <uri>https://example.test/audio.mp3</uri>
    </resource>
  </info>
</alert>
"#;
        let alert = parse_cap(raw).expect("cap parsed");
        let value = serde_json::to_value(&alert).expect("json");
        assert_eq!(value["identifier"], "abc");
        assert_eq!(value["message_type"], "Alert");
        assert_eq!(value["infos"][0]["response_type"], json!(["Shelter"]));
        assert_eq!(value["infos"][0]["event_codes"][0]["value"], "SVR");
        assert_eq!(
            value["infos"][0]["areas"][0]["geocodes"][0]["value"],
            "065100"
        );
    }

    #[test]
    fn parses_atom_links_and_cap_fallbacks() {
        let raw = br#"
<feed xmlns="http://www.w3.org/2005/Atom">
  <entry>
    <id>https://alerts.example/item/1</id>
    <updated>2026-06-22T12:39:00Z</updated>
    <link href="https://alerts.example/item/1"/>
  </entry>
</feed>
"#;
        let entries = parse_atom_entries(raw).expect("atom parsed");
        assert_eq!(entries.len(), 1);
        assert_eq!(entries[0].id, "https://alerts.example/item/1");
        assert_eq!(entries[0].links[0], "https://alerts.example/item/1");
        assert_eq!(entries[0].links[1], "https://alerts.example/item/1.cap");
    }

    #[test]
    fn accepts_naads_heartbeat_without_info() {
        let raw = br#"
<alert xmlns="urn:oasis:names:tc:emergency:cap:1.2">
  <identifier>heartbeat-1</identifier>
  <sender>NAADS-Heartbeat</sender>
  <sent>2026-06-22T12:38:00-06:00</sent>
  <status>System</status>
  <msgType>Alert</msgType>
  <scope>Public</scope>
  <references>sender,abc,2026-06-22T12:37:00-06:00</references>
</alert>
"#;
        let alert = parse_cap(raw).expect("heartbeat parsed");
        assert_eq!(alert.status, "System");
        assert!(alert.infos.is_empty());
    }
}
